### PR TITLE
THRIFT-5544: add java code gen param to support including field annotation as metadata

### PR DESCRIPTION
**Pull request by Jimexist**
add java code gen param to support including field annotation as metadata

Currently the code generator has a map of string to string field that denotes the "annotations" for a field: https://github.com/apache/thrift/blob/90ea2e8398eda32da8be0b3514516e7ad932a869/compiler/cpp/src/thrift/parse/t_field.h#L109

It is made available to generators such as JSON (in https://github.com/apache/thrift/blob/90ea2e8398eda32da8be0b3514516e7ad932a869/compiler/cpp/src/thrift/generate/t_json_generator.cc#L267-L274) to allow (runtime) interpretation of such metadata.

In Java, similar mechanism exists to register field metadata (via https://github.com/apache/thrift/blob/90ea2e8398eda32da8be0b3514516e7ad932a869/compiler/cpp/src/thrift/generate/t_java_generator.cc#L2862-L2865 which calls `org.apache.thrift.meta_data.FieldMetaData#addStructMetaDataMap` method to register into a centralized places. Such metadata is useful, e.g. the microservice framework Ameria [uses it][1] to dynamically proxy requests.

However the existing field metadata only contain information about field id, type, and name, but it does not contain annotations. Adding annotations will be useful for situations e.g. (these are useful enablement at user side, make possible if annotation is included)

- some static metadata is useful at runtime for dynamic operations, e.g. like what `deprecated` means in the context of a field, adding `obfuscated` allows the user site to decide to obfuscate detailed information
- allows for more richer and more type specific validation and transformation logic, e.g. since we can only have string as field type, adding `date_format="YYYY-MM-DD"` allows the thrift struct to be interpreted as a local date and thus can automatically be converted to java.time.LocalDate

This should be a back-compatible change as the default behavior is not to include the annotation. A second constructor can be added to allow included annotation to exist and register.

- [x] Did you create an [Apache Jira](https://issues.apache.org/jira/projects/THRIFT/issues/) ticket?  (not required for trivial changes)
- [x] If a ticket exists: Does your pull request title follow the pattern "THRIFT-NNNN: describe my issue"?
- [x] Did you squash your changes to a single commit?  (not required, but preferred)
- [x] Did you do your best to avoid breaking changes?  If one was needed, did you label the Jira ticket with "Breaking-Change"?
- [x] If your change does not involve any code, include `[skip ci]` anywhere in the commit message to free up build resources.

[1]: https://github.com/line/armeria/blob/5c2efebc7ba2350540d0554f530c173a49eb4c19/thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpService.java#L553-L554

### CODE REVIEW

**Comment by Jens-G** [comment]
>  but a quick check shows it's defaults would change a lot of thrift code, and not always for the better. 
> I'm not sure if the Thrift community would endorse its use,

There were similar approaches a while ago and my (personal) feelings about automated C++ code formatting tools abilities are, well, let's call it "mixed feelings". But it's not only a C++ specific thing, I had the same experience with other tools and other languages as well. Sometimes the problems start with the style guide supplied by the manufacturer itself that ist a) the default and b) horribly broken. (Yes, I'm looking at you, Emba). 

The best tool I've seen in that regard is surely `gofmt`. The style actually makes sense to begin with, and it also offers not a single formatting-related option that could be configured. None. Zero. Kind of "get some sane style without the bikeshedding" approach.

**Comment by Jimexist** [comment]
> > but a quick check shows it's defaults would change a lot of thrift code, and not always for the better.
> > I'm not sure if the Thrift community would endorse its use,
> 
> There were similar approaches a while ago and my (personal) feelings about automated C++ code formatting tools abilities are, well, let's call it "mixed feelings". But it's not only a C++ specific thing, I had the same experience with other tools and other languages as well. Sometimes the problems start with the style guide supplied by the manufacturer itself that ist a) the default and b) horribly broken. (Yes, I'm looking at you, Emba).
> 
> The best tool I've seen in that regard is surely `gofmt`. The style actually makes sense to begin with, and it also offers not a single formatting-related option that could be configured. None. Zero. Kind of "get some sane style without the bikeshedding" approach.

@Jens-G did you mean to comment on https://github.com/apache/thrift/pull/2559#pullrequestreview-938818662 instead?

**Comment by Jimexist** [comment]
> 

I'd say that both [rustfmt](https://github.com/rust-lang/rustfmt) and gofmt are in this category and it contribute to the fact that Rust's one of the most love languages (per stackoverflow)

**Comment by Jens-G** [comment]
> did you mean to comment on [...]

Probably. Must have confused open tabs.

**Review by ctubbsii — COMMENTED** [review]
I think this could use a larger explanation. From these changes, including the added help documentation in the generator options, it's not clear what 'annotation metadata' is. What are common examples of annotation metadata, and how would these be used? The JIRA doesn't have any details in it, either: https://issues.apache.org/jira/browse/THRIFT-5544

**Comment by Jimexist** [comment]
> I think this could use a larger explanation. From these changes, including the added help documentation in the generator options, it's not clear what 'annotation metadata' is. What are common examples of annotation metadata, and how would these be used? The JIRA doesn't have any details in it, either: https://issues.apache.org/jira/browse/THRIFT-5544

@ctubbsii thanks for the comment. detailed explanation added to jira ticket and inline

**Comment by Jimexist** [comment]
> > > I think this could use a larger explanation. From these changes, including the added help documentation in the generator options, it's not clear what 'annotation metadata' is. What are common examples of annotation metadata, and how would these be used? The JIRA doesn't have any details in it, either: https://issues.apache.org/jira/browse/THRIFT-5544
> > 
> > 
> > @ctubbsii thanks for the comment. detailed explanation added to jira ticket and inline
> 
> Thanks for the additional information. However, I'm still confused. Is this intended to serialize information about Java annotations `@Override`, `@Deprecated`, `@Nullable`, etc.? Are these annotations in the Java code that is generated from an `*.thrift` IDL file? Is the annotation defined in the `*.thrift` file or expected to be manually added after code generation?
> 
> I'm just not clear on the end-to-end use of this. Do you have a simple example?

@ctubbsii i created a simple example at https://gist.github.com/Jimexist/7acffaa0c4967193937a05f56098853b with the newly added flag in this pull request.

**Comment by ctubbsii** [comment]
> @ctubbsii i created a simple example at https://gist.github.com/Jimexist/7acffaa0c4967193937a05f56098853b with the newly added flag in this pull request.

Thanks. That example helps clear up some things. I was confused by the use of the terms "java" and "field annotation", because I was thinking [Java's Field `@Annotations`](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/Field.html#getAnnotation-java.lang.Class-).

I would call these "tags" rather than "annotations", in order to reduce confusion. However, if this is already an existing IDL language construct, then the name may already be established as "field annotations". If that's the case, I can propose some slight wording improvements in the gen-java parameter help doc.

File: compiler/cpp/src/thrift/generate/t_java_generator.cc [hunk 1]
```diff
@@ -78,6 +78,7 @@ class t_java_generator : public t_oop_generator {
     suppress_generated_annotations_ = false;
     rethrow_unhandled_exceptions_ = false;
     unsafe_binaries_ = false;
+    annotation_metadata_ = false;
```
Comment (ctubbsii) [inline compiler/cpp/src/thrift/generate/t_java_generator.cc:81]: Maybe:
```suggestion
    annotations_as_metadata_ = false;
```

File: compiler/cpp/src/thrift/generate/t_java_generator.cc [hunk 2]
```diff
@@ -2922,6 +2932,33 @@ std::string t_java_generator::get_java_type_string(t_type* type) {
   }
 }
 
+void t_java_generator::generate_field_annotations(std::ostream& out, t_field* field) {
```
Comment (ctubbsii) [inline compiler/cpp/src/thrift/generate/t_java_generator.cc:2935]:
```suggestion
void t_java_generator::generate_metadata_for_field_annotations(std::ostream& out, t_field* field) {
```

File: compiler/cpp/src/thrift/generate/t_java_generator.cc [hunk 3]
```diff
@@ -5737,4 +5774,6 @@ THRIFT_REGISTER_GENERATOR(
     "    generated_annotations=[undated|suppress]:\n"
     "                     undated: suppress the date at @Generated annotations\n"
     "                     suppress: suppress @Generated annotations entirely\n"
-    "    unsafe_binaries: Do not copy ByteBuffers in constructors, getters, and setters.\n")
+    "    unsafe_binaries: Do not copy ByteBuffers in constructors, getters, and setters.\n"
+    "    annotation_metadata:\n"
+    "                     Include annotation metadata in the generated code.\n")
```
Comment (ctubbsii) [inline compiler/cpp/src/thrift/generate/t_java_generator.cc:5779]:
```suggestion
    "    annotations_as_metadata:\n"
    "                     Include annotations as metadata in the generated code.\n")
```

File: lib/java/gradle/generateTestThrift.gradle [hunk 4]
```diff
@@ -133,3 +134,12 @@ task generateUnsafeBinariesJava(group: 'Build') {
 
     thriftCompile(it, 'UnsafeTypes.thrift', 'java:unsafe_binaries', genUnsafeSrc)
 }
+
+task generateWithAnnotation(group: 'Build') {
```
Comment (ctubbsii) [inline lib/java/gradle/generateTestThrift.gradle:138]:
```suggestion
task generateWithAnnotationMetadata(group: 'Build') {
```

**Review by ctubbsii — COMMENTED** [review]
Here's some proposed naming / phrasing changes to clear up potential confusion.

**Comment by Jimexist** [comment]
> However, if this is already an existing IDL language construct, then the name may already be established as "field annotations".



> > @ctubbsii i created a simple example at https://gist.github.com/Jimexist/7acffaa0c4967193937a05f56098853b with the newly added flag in this pull request.
> 
> Thanks. That example helps clear up some things. I was confused by the use of the terms "java" and "field annotation", because I was thinking [Java's Field `@Annotations`](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/Field.html#getAnnotation-java.lang.Class-).
> 
> I would call these "tags" rather than "annotations", in order to reduce confusion. However, if this is already an existing IDL language construct, then the name may already be established as "field annotations". If that's the case, I can propose some slight wording improvements in the gen-java parameter help doc.

I agree it's confusing but i guess that ship is sailed: the name in C++ and also in all of the file names it's named as "annotation". I guess the best method is to try best to make it less ambiguous in Java.

**Comment by Jimexist** [comment]
> However, if this is already an existing IDL language construct, then the name may already be established as "field annotations".



> > @ctubbsii i created a simple example at https://gist.github.com/Jimexist/7acffaa0c4967193937a05f56098853b with the newly added flag in this pull request.
> 
> Thanks. That example helps clear up some things. I was confused by the use of the terms "java" and "field annotation", because I was thinking [Java's Field `@Annotations`](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/Field.html#getAnnotation-java.lang.Class-).
> 
> I would call these "tags" rather than "annotations", in order to reduce confusion. However, if this is already an existing IDL language construct, then the name may already be established as "field annotations". If that's the case, I can propose some slight wording improvements in the gen-java parameter help doc.

I agree it's confusing but i guess that ship is sailed: the name in C++ and also in all of the file names it's named as "annotation". I guess the best method is to try best to make it less ambiguous in Java.

File: compiler/cpp/src/thrift/generate/t_java_generator.cc [hunk 5]
```diff
@@ -5737,4 +5774,6 @@ THRIFT_REGISTER_GENERATOR(
     "    generated_annotations=[undated|suppress]:\n"
     "                     undated: suppress the date at @Generated annotations\n"
     "                     suppress: suppress @Generated annotations entirely\n"
-    "    unsafe_binaries: Do not copy ByteBuffers in constructors, getters, and setters.\n")
+    "    unsafe_binaries: Do not copy ByteBuffers in constructors, getters, and setters.\n"
+    "    annotations_as_metadata:\n"
+    "                     Include annotations as metadata in the generated code.\n")
```
Comment (ctubbsii) [inline compiler/cpp/src/thrift/generate/t_java_generator.cc:5779]: Just a tiny bit more clarification here would help I think.

```suggestion
    "                     Include Thrift field annotations as metadata in the generated code.\n")
```

**Review by ctubbsii — COMMENTED** [review]
I have only this last final suggestion to help further clarify things in the documentation. But otherwise, this looks great to me.

**Review by ctubbsii — APPROVED** [review]

**Comment by Jimexist** [comment]
had to do a rebase onto master due to conflict.

**Comment by Jimexist** [comment]
seems like the build failure was due to `kerl` failing to compile erlang in time...

**Comment by Jimexist** [comment]
now this is fixed, ready for merging

**Pull request merged**